File: course1/week2/2_last_digit_of_fibonacci_number/java/src/Main.java
/**
 *
 * Java implementation to find the last digit of a large fibonacci number
 *
 * (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
 *
 **/

import java.util.ArrayList;
import java.util.Scanner;
import java.util.Collections;

public class Main {

    public static Integer INF = Integer.MAX_VALUE;

    public static Integer fibonacci( Integer N ){
        if( N < 2 )
            return N;
        ArrayList< Integer > dp = new ArrayList<>( Collections.nCopies( N+1, INF ));
        dp.set( 0, 0 );
        dp.set( 1, 1 );
        for( int i = 2; i <= N; ++i )
            dp.set( i, (dp.get( i-2 ) + dp.get( i-1 )) % 10 );
        return dp.get( N );
    }

    public static void main( String[] args ){
        Scanner input = new Scanner( System.in );
        Integer N = input.nextInt(),
                ans = fibonacci( N );
        System.out.println( ans );
    }
}

File: course1/week2/2_last_digit_of_fibonacci_number/python/main.py
# python3

##
#
# Python implementation to find the last digit of a large fibonacci number
#
# (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
#
##

INF = 999999999

def fibonacci( N ):
    if N < 2:
        return N
    dp = ( N+1 ) * [ INF ]
    dp[ 0 ] = 0
    dp[ 1 ] = 1
    for i in range( 2, N+1 ):
        dp[ i ] = ( dp[ i-2 ] + dp[ i-1 ] ) % 10
    return dp[ N ]

if __name__ == "__main__":
    N = int( input() )
    ans = fibonacci( N )
    print( ans )
